### PR TITLE
Announcements - make announcements directly from events, meetings and interest groups

### PR DESCRIPTION
**Pull request by liseph**
Text type:
image

Button type:
image 1

When clicked:
image 2
image 3

Meetings:
screenshot from 2017-11-02 23-08-00

Interest groups:
screenshot from 2017-11-02 23-09-23

Events:
screenshot from 2017-11-02 23-10-05

### CODE REVIEW

File: lib/state_example.js [hunk 1]
```diff
@@ -0,0 +1,1397 @@
+module.exports = {
+  location: {
+    pathname: '/events',
+    search: '',
```
Comment (finnss) [inline lib/state_example.js]: I'm guessing this file should be in `.gitignore`?

File: yarn.lock [hunk 2]
```diff
@@ -1,33 +1,24 @@
 # THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
 # yarn lockfile v1
-
-
 "@types/node@^6.0.46":
```
Comment (finnss) [inline yarn.lock]: I'm not an expert on `yarn.lock`, but 1000 lines changed seems excessive?

File: app/components/ProfilePicture/Readme.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Example profilePicture:
+```
+<div>
+  <ProfilePicture user={state.currentUser} />
+  <p> {state.currentUser.fullName} </p>
```
Comment (finnss) [inline app/components/ProfilePicture/Readme.md]: Guessing this isn't intended to be included in this PR?

File: app/components/Search/Readme.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+```
+const { Provider } = require('react-redux');
+const configureStore = require('app/utils/configureStore').default;
+const store = configureStore({ state });
```
Comment (finnss) [inline app/components/Search/Readme.md]: Guessing this isn't intended to be included in this PR?

File: app/routes/announcements/components/AnnouncementItem.js [hunk 5]
```diff
@@ -84,19 +84,22 @@ const AnnouncementItem = ({
         </Flex>
       </Flex>
       {!announcement.sent &&
-        actionGrant.includes('send') && (
+        actionGrant.includes('send') &&
+        actionGrant.includes('delete') && (
```
Comment (finnss) [inline app/routes/announcements/components/AnnouncementItem.js]: Is there a difference between who can send and who can delete an announcement?

File: app/routes/announcements/components/AnnouncementItem.js [hunk 6]
```diff
@@ -84,19 +84,22 @@ const AnnouncementItem = ({
...
           <Flex className={styles.wrapperSendButton}>
             <Button
               className={styles.sendButton}
               onClick={() => deleteAnnouncement(announcement.id)}
             >
-              Slett
+              {' '}
+              Slett{' '}
```
Comment (finnss) [inline app/routes/announcements/components/AnnouncementItem.js]: Are these spaces necessary? I think they're kinda fine to use in the middle of a sentence where setting a margin just to imitate a space would be akward, like 
```
Han er{' '}
{longFunctionDetermeningHowOldTheUserIs()}
{' '}år
```
but they aren't a good way to imitate margin or padding.

File: app/routes/announcements/components/AnnouncementsCreate.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import React from 'react';
+import styles from './AnnouncementsList.css';
+import Flex from 'app/components/Layout/Flex';
+import { Form, SelectInput, TextArea } from 'app/components/Form';
+import { reduxForm, Field, reset } from 'redux-form';
+import Button from 'app/components/Button';
+import { ContentMain } from 'app/components/Content';
+
+type Props = {
+  createAnnouncement: (...any) => void,
```
Comment (finnss) [inline app/routes/announcements/components/AnnouncementsCreate.js]: Would be best if you made an `AnnouncementEntity` that determines what fields an announcement should have. Probably need one for displaying them and one to determine what fields are sent when creating one, like `CreateAnnouncementEntity` or some such. See reducers for examples. Also, this function probably returns `Promise<*>`, not void.

File: app/components/AnnouncementInLine/index.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+import { reduxForm, Field, reset } from 'redux-form';
+import { createAnnouncement } from 'app/actions/AnnouncementsActions';
+import styles from './AnnouncementInLine.css';
+
+type Props = {
+  placeholder?: string,
+  event?: number,
+  meeting?: number,
+  group?: number,
+  createAnnouncement: (...any) => void,
```
Comment (finnss) [inline app/components/AnnouncementInLine/index.js]: See other comment about this

File: app/routes/announcements/components/AnnouncementsCreate.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        ? announcement.meetings.map(meeting => meeting.value)
+        : [],
+      events: announcement.events
+        ? announcement.events.map(event => event.value)
+        : [],
+      send
+    });
+  };
+  return (
+    <ContentMain>
```
Comment (finnss) [inline app/routes/announcements/components/AnnouncementsCreate.js]: @ekmartin would be proud

File: app/routes/announcements/components/AnnouncementsList.js [hunk 10]
```diff
@@ -1,18 +1,15 @@
...
-import { reduxForm, Field } from 'redux-form';
-import Button from 'app/components/Button';
 import AnnouncementItem from './AnnouncementItem';
+import AnnouncementsCreate from './AnnouncementsCreate';
+import { Content, ContentMain } from 'app/components/Content';
 
 type Props = {
   announcement: Object,
-  announcements: Array</*TODO: Announcement */ any>,
+  announcements: Array<Object>,
```
Comment (finnss) [inline app/routes/announcements/components/AnnouncementsList.js]: These should both be `AnnouncementEntity`, see other comment.

File: app/routes/announcements/components/AnnouncementsCreate.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+}: Props) => {
+  const disabledButton = invalid || pristine || submitting;
+
+  const onSubmit = (announcement, send = false) => {
+    return createAnnouncement({
+      ...announcement,
+      users: announcement.users
+        ? announcement.users.map(user => user.value)
+        : [],
+      groups: announcement.groups
```
Comment (finnss) [inline app/routes/announcements/components/AnnouncementsCreate.js]: You could probably avoid all these checks if you passed AnnouncementCreate some `initialValues`:

```
const AnnouncementsCreate = ...;

const initialValues = {
    events: [],
    meetings: [],
    ...
}

export default reduxForm({
   initialValues,
   ...
})(AnnouncementCreate);
```

File: app/routes/announcements/components/AnnouncementsList.js [hunk 12]
```diff
@@ -1,18 +1,15 @@
...
   actionGrant: /* TODO: ActionGrant */ any,
   sendAnnouncement: () => void,
   createAnnouncement: (announcement: any) => void,
```
Comment (finnss) [inline app/routes/announcements/components/AnnouncementsList.js]: These both probably return promises. announcement should be an entity, not any.

File: app/routes/announcements/components/AnnouncementsCreate.js (hunk 7 above)
Comment (odinuge) [reply app/routes/announcements/components/AnnouncementsCreate.js]: Create a model for it here: https://github.com/webkom/lego-webapp/blob/master/app/models.js

File: app/components/Search/Readme.md (hunk 4 above)
Comment (odinuge) [reply app/components/Search/Readme.md]: This was removed upstream, and should be removed. The `state-example` should also be removed from the commit it was introduced.

File: app/routes/announcements/components/AnnouncementsCreate.js (hunk 7 above)
Comment (finnss) [reply app/routes/announcements/components/AnnouncementsCreate.js]: I guess I need to move all of my Entities there.. 😅

**Pull request merged**